src/tactic/lift.lean

```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2019 Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn
+-/
+import tactic.rcases
+
+/-!
+# lift tactic
+
+This file defines the lift tactic, allowing the user to lift elements from one type to another
+under a specified condition.
+
+## Tags
+
+lift, tactic
+-/
+
+universe variables u v
+
+run_cmd mk_simp_attr `can_lift
+
+/-- A class specifying that you can lift elements from `α` to `β` assuming `cond` is true.
+  Used by the tactic `lift`. -/
+class can_lift (α : Type u) (β : Type v) : Type (max u v) :=
```

Comment:
You could declare this class as `class can_lift (α : Type u) (β : Type v) [has_coe α β] : Type (max u v) :=` so that you can use the standard coercion mechanism instead of `coe` declared in this class.

Reply:
What if we want to use the lift tactic in cases where we don't want a coercion from `β` to `α`? 
I think the lift tactic could be more general. 

For example: you add an instance of `can_lift (α ⊕ β) α` if you have a proof that `∀y, (x : α ⊕ β) ≠ sum.inr y`. Doing this with `lift` is probably not really easier than just doing cases on `x`, but it is possible. And there are probably situations where it is actually useful. Maybe you have a type that is equivalent to `α ⊕ β` (or maybe you just have a map from `α ⊕ β`), where a `can_lift` instance would be easier than going along the equivalence yourself.

This tactic might also occasionally be useful when you have an equivalence `α ≃ β` and want to move elements between `α` and `β`.

Reply:
That's fair. I did not consider those use cases. In your examples, `can_lift.coe` seems to disappear` after using `lift`. Does your tactic make sure of that?

Reply:
Yes, I run `dsimp only with can_lift` on all resulting goals/hypotheses (and even in some error messages) - search the tactic for `dsimp`. That's why the `can_lift` instance has the `can_lift` attribute.
```
@[can_lift] instance : can_lift ℤ ℕ
```

Reply:
Why not just by default include every `can_lift` instance in that `simp` set?

Reply:
how do I do that?

Reply:
`simp` is a bit tricky to script. What I would do is use `simp_target` or `simp_hyp` (depending on what you're trying to rewrite) and use `mk_simp_set` to construct the `simp_lemmas` argument. When you call `mk_simp_set`, you give it a list of `simp_arg_type` built from the `pexpr` of all the instances of `can_lift`

Reply:
And I get all these instances by using ```attribute.get_instances `instance```, and for all of them looking at the target of the type? 
Is that efficient enough? That seems like it could be a performance hit for the very minor convenience of not having to write `@[can_lift]`. I'm leaning towards just using `@[can_lift]`.

Reply:
You can index those instances with this kind of code:

```lean
open tactic

@[user_attribute]
meta def foo : user_attribute (list name) := 
{ name := "a",
  descr := "d",
  cache_cfg := { mk_cache := λ _,
    do{ ls ← attribute.get_instances `instance,
        ls.mfilter $ λ l, 
        do { (_,t) ← mk_const l >>= infer_type >>= mk_local_pis, 
         return $ t.is_app_of `decidable_eq } },
  dependencies := [`instance] } }

-- @[a]
instance decidable_eq.foo : decidable_eq ℕ := sorry

run_cmd do 
l ← foo.get_cache,
trace l.length
```

Reply:
That code nominally declares an attribute that you never have to use. The attribute only serves as a location for an instance cache. The cache is rebuilt every time you declare an instance and the cache_cfg code keeps only `decidable_eq` instances (or `can_lift` in your case).

Reply:
Ah, that is pretty nice. I didn't think of using the attribute cache for this. Thank you for writing the `cache_cfg` for this attribute.

I now used this code (and added another test checking that it works properly).

Reply:
I agree, this is clever!

Reply:
I'm pretty happy with it myself. I only recently managed to understand what `dependencies` does.